addons/stock/models/stock_move.py

```diff
@@ -874,18 +875,24 @@ def _merge_moves(self, merge_into=False):
         for neg_move in neg_qty_moves:
             # Check all the candidates that matches the same limited key, and adjust their quantites to absorb negative moves
             for pos_move in moves_by_neg_key.get(neg_key(neg_move), []):
-                # If quantity can be fully absorbed by a single move, update its quantity and remove the negative move
-                if float_compare(pos_move.product_uom_qty, abs(neg_move.product_uom_qty), precision_rounding=pos_move.product_uom.rounding) >= 0:
-                    pos_move.product_uom_qty += neg_move.product_uom_qty
-                    pos_move.write({
-                        'move_dest_ids': [Command.link(m.id) for m in neg_move.mapped('move_dest_ids') if m.location_id == pos_move.location_dest_id],
-                        'move_orig_ids': [Command.link(m.id) for m in neg_move.mapped('move_orig_ids') if m.location_dest_id == pos_move.location_id],
-                    })
-                    merged_moves |= pos_move
-                    moves_to_unlink |= neg_move
-                    break
-                neg_move.product_uom_qty += pos_move.product_uom_qty
-                pos_move.product_uom_qty = 0
+                currency_prec = pos_move.product_id.currency_id.decimal_places
+                rounding = min(currency_prec, price_unit_prec)
+                if float_compare(pos_move.price_unit, neg_move.price_unit, precision_rounding=rounding) == 0:
```

Comment:
I don't get the reason why we would not merge two moves with a different price unit ? 
Was there an issue with the simple diff I have suggested in my first review ? (cc @Whenrow )
Otherwise, you can remove the price unit computation (line 894) as it would become useless

Reply:
I just assumed that if we buy something for 10$ for exemple and the vendor only accepts to return for 6$ the two moves should be kept

Reply:
The cost of the product is the one on the purchase order line. If the positive move and the negative one share the same PO line, their cost should be same and be merged IMO